mod/quiz/accessrule/seb/classes/access_manager.php

```diff
@@ -73,48 +73,105 @@ public function __construct(quiz $quiz) {
+     * Validate SEB keys that should be checked. If the keys are valid, store the access state in Moodle session.
+     *
+     * @param string|null $configkey Optionally provide a SEB config key to validate.
+     * @param string|null $browserexamkey Optionally provide a SEB browser exam key to validate.
+     * @param string|null $url Optionally provide URL of page to validate.
+     * @return bool
+     */
+    public function validate_seb_keys(?string $configkey = null, ?string $browserexamkey = null, ?string $url = null): bool {
+        // Check if there is a valid config key.
+        if ($this->should_validate_config_key() && !$this->validate_config_key($configkey, $url)) {
+            return false;
+        }
+
+        // Check if there is a valid browser exam key.
+        if ($this->should_validate_browser_exam_key() && !$this->validate_browser_exam_key($browserexamkey, $url)) {
+            return false;
+        }
+
+        // Set the state of the access for this Moodle session.
+        $this->set_session_access(true);
+
+        return true;
+    }
+
+    /**
+     * Validate browser exam key. It will validate a provided browser exam key if provided, then will fall back to checking
+     * the header.
      *
-     * @return bool True if header key matches one of the saved keys.
+     * @param string|null $browserexamkey Optional. Can validate a provided key, or will fall back to checking header.
+     * @return bool
      */
-    public function validate_browser_exam_keys() : bool {
-        // If browser exam keys are entered in settings, check they match the header.
-        $browserexamkeys = $this->quizsettings->get('allowedbrowserexamkeys');
-        if (empty($browserexamkeys)) {
+    public function validate_browser_exam_key(?string $browserexamkey = null, ?string $url = null): bool {
+        if (!$this->should_validate_browser_exam_key()) {
+            throw new \coding_exception('Check whether the browser exam key should be validated with $accessmanager->should_validate_browser_exam_key()');
+        }
+
+        if (!$this->valid_browser_exam_key_exists()) {
             return true; // If no browser exam keys, no check required.
         }
 
+        if (empty($browserexamkey)) {
+            $browserexamkey = $this->get_received_browser_exam_key();
+        }
+
+        $validbrowserexamkeys = $this->quizsettings->get('allowedbrowserexamkeys');
+
         // If the Browser Exam Key header isn't present, prevent access.
-        if (is_null($this->get_received_browser_exam_key())) {
+        if (is_null($browserexamkey)) {
             return false;
         }
 
-        return $this->check_browser_exam_keys($browserexamkeys, $this->get_received_browser_exam_key());
+        $valid = $this->check_browser_exam_keys($validbrowserexamkeys, $browserexamkey, $url);
+
+        if (!$valid) {
+            access_prevented::create_strict($this, $this->get_reason_text('invalid_browser_key'))->trigger();
```

Comment:
Those events shouldn't be in this class. They should be closer to where we actually prevent the access. That may require refactoring web service and get rid of validate_seb_keys method to make it look like the original prevent_access method.